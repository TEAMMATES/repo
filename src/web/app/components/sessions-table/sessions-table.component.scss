.no-session-message {
  padding: 10px;
  text-align: center;
}

@media (min-width: 1200px) {
  .actions-cell {
    min-width: 410px;
  }
}

.actions-cell button:not(.dropdown-item) {
  margin-left: var(--btn-margin);
}

.sortable-header {
  cursor: pointer;
}

.clickable {
  cursor: pointer;
}

.margin-bottom-0 {
  margin-bottom: 0;
}

.align-row {
  display: flex;
  flex-direction: row;
}

.loading {
  margin-left: 6px;
  margin-top: 3px;
  width: 12px;
  height: 12px;
  border: 2px solid transparent;
  border-top-color: black;
  border-radius: 50%;
  animation: loading-spinner 1s ease infinite;
}

@keyframes loading-spinner {
  from {
    transform: rotate(0turn);
  }

  to {
    transform: rotate(1turn);
  }
}
