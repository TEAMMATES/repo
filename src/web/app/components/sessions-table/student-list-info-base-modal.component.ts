import { NgbActiveModal } from '@ng-bootstrap/ng-bootstrap';
import { HttpRequestService } from '../../../services/http-request.service';
import { StatusMessageService } from '../../../services/status-message.service';
import { ErrorMessageOutput } from '../../error-message-output';
import { StudentFeedbackSessionResponseStatus, StudentsFeedbackSessionResponseStatus } from '../../student';
import { SortBy, SortOrder, StudentStatusTableRowModel } from './sessions-table-model';

/**
 * The base modal information for a list of students.
 */
export abstract class StudentListInfoBaseModalComponent {

  // enum
  protected SortBy: typeof SortBy = SortBy;
  protected SortOrder: typeof SortOrder = SortOrder;

  protected studentStatusTableRows: StudentStatusTableRowModel[] = [];
  protected studentsTableRowSortBy: SortBy = SortBy.NONE;
  protected studentsTableRowSortOrder: SortOrder = SortOrder.DESC;

  protected loading: boolean = false;
  protected isAjaxSuccess: boolean = true;

  protected constructor(protected activeModal: NgbActiveModal, protected httpRequestService: HttpRequestService,
                        protected statusMessageService: StatusMessageService) { }

  /**
   * Get the list of students to remind in table form.
   */
  getStudentStatusTableRowModel(paramMap: { [key: string]: string }, model: StudentStatusTableRowModel[]): void {
    this.httpRequestService.get('/session/remind/submission', paramMap)
      .subscribe((studentsFeedbackSessionResponseStatus: StudentsFeedbackSessionResponseStatus) => {
        studentsFeedbackSessionResponseStatus.studentsFeedbackSessionResponseStatus
          .forEach((studentFeedbackSessionResponseStatus: StudentFeedbackSessionResponseStatus) => {

            const studentStatusTableRowModel: StudentStatusTableRowModel = {
              studentFeedbackSessionResponseStatus,
              isChecked: false,
            };

            model.push(studentStatusTableRowModel);
          });
        this.sortStudentsTableRows(SortBy.SUBMIT_STATUS);
      }, (resp: ErrorMessageOutput) => {
        this.statusMessageService.showErrorMessage(resp.error.message);
        this.isAjaxSuccess = false;
      });
    this.loading = false;
  }

  /**
   * Sort the students in according to selection option.
   */
  sortStudentsTableRows(by: SortBy): void {
    this.studentsTableRowSortBy = by;
    // reverse the sort order
    this.studentsTableRowSortOrder =
        this.studentsTableRowSortOrder === SortOrder.DESC ? SortOrder.ASC : SortOrder.DESC;
    this.studentStatusTableRows.sort(this.sortRowsBy(by, this.studentsTableRowSortOrder));
  }

  /**
   * Sorts the rows of students in order.
   */
  sortRowsBy(by: SortBy, order: SortOrder):
      ((a: { studentFeedbackSessionResponseStatus: StudentFeedbackSessionResponseStatus },
        b: { studentFeedbackSessionResponseStatus: StudentFeedbackSessionResponseStatus }) => number) {
    return ((a: { studentFeedbackSessionResponseStatus: StudentFeedbackSessionResponseStatus },
             b: { studentFeedbackSessionResponseStatus: StudentFeedbackSessionResponseStatus }): number => {
      let strA: string;
      let strB: string;
      switch (by) {
        case SortBy.SECTION_NAME:
          strA = a.studentFeedbackSessionResponseStatus.sectionName;
          strB = b.studentFeedbackSessionResponseStatus.sectionName;
          break;
        case SortBy.TEAM_NAME:
          strA = a.studentFeedbackSessionResponseStatus.teamName;
          strB = b.studentFeedbackSessionResponseStatus.teamName;
          break;
        case SortBy.STUDENT_NAME:
          strA = a.studentFeedbackSessionResponseStatus.name;
          strB = b.studentFeedbackSessionResponseStatus.name;
          break;
        case SortBy.STUDENT_EMAIL:
          strA = a.studentFeedbackSessionResponseStatus.email;
          strB = b.studentFeedbackSessionResponseStatus.email;
          break;
        case SortBy.SUBMIT_STATUS:
          strA = a.studentFeedbackSessionResponseStatus.responseStatus.toString();
          strB = b.studentFeedbackSessionResponseStatus.responseStatus.toString();
          break;
        default:
          strA = '';
          strB = '';
      }
      if (order === SortOrder.ASC) {
        return strA.localeCompare(strB);
      }
      if (order === SortOrder.DESC) {
        return strB.localeCompare(strA);
      }
      return 0;
    });
  }

  /**
   * Check all the checkbox of the students.
   */
  checkAllStudents(model: StudentStatusTableRowModel[], isCheck: boolean): void {
    for (const remindStudentRow of model) {
      remindStudentRow.isChecked = isCheck;
    }
  }

  /**
   * Collates a list of selected students with selected checkbox.
   */
  collateStudentsToSend(model: StudentStatusTableRowModel[]): string[] {
    const remindStudentList: string[] = [];
    for (const studentStatusTableRow of model) {
      if (studentStatusTableRow.isChecked) {
        remindStudentList.push(studentStatusTableRow.studentFeedbackSessionResponseStatus.email);
      }
    }
    return remindStudentList;
  }
}
