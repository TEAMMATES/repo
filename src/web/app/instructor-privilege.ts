/**
 * Instructor privilege.
 */
export interface InstructorPrivilege {
  canModifyCourse: boolean;
  canModifySession: boolean;
  canModifyStudent: boolean;
  canSubmitSessionInSections: boolean;
}

/**
 * The default instructor privilege.
 */
export const defaultInstructorPrivilege: InstructorPrivilege = {
  canModifyCourse: true,
  canModifySession: true,
  canModifyStudent: true,
  canSubmitSessionInSections: true,
};

/**
 * The format for instructor privileges.
 */
export interface Privileges {
  courseLevel: CourseLevelPrivileges;

  // Maps a section name to section level privileges
  sectionLevel: { [section: string]: SectionLevelPrivileges };

  // Maps a section name to a map mapping the session name and session privileges
  sessionLevel: { [section: string]: { [session: string]: SessionLevelPrivileges } };
}

/**
 * The format for instructor course privileges.
 */
export interface CourseLevelPrivileges {
  canmodifycourse: boolean;
  canmodifyinstructor: boolean;
  canmodifysession: boolean;
  canmodifystudent: boolean;
  canviewstudentinsection: boolean;
  canviewsessioninsection: boolean;
  cansubmitsessioninsection: boolean;
  canmodifysessioncommentinsection: boolean;
  [key: string]: boolean;
}

/**
 * The format for instructor section privileges.
 */
export interface SectionLevelPrivileges {
  canviewstudentinsection: boolean;
  canviewsessioninsection: boolean;
  cansubmitsessioninsection: boolean;
  canmodifysessioncommentinsection: boolean;
  [key: string]: boolean;
}

/**
 * The format for instructor session privileges.
 */
export interface SessionLevelPrivileges {
  canviewsessioninsection: boolean;
  cansubmitsessioninsection: boolean;
  canmodifysessioncommentinsection: boolean;
  [key: string]: boolean;
}

/**
 * Possible instructor roles.
 */
export enum Role {
  /**
   * Co-owner instructor role.
   */
  COOWNER = 'Co-owner',

  /**
   * Manager instructor role.
   */
  MANAGER = 'Manager',

  /**
   * Observer instructor role.
   */
  OBSERVER = 'Observer',

  /**
   * Tutor instructor role.
   */
  TUTOR = 'Tutor',

  /**
   * Custom instructor role.
   */
  CUSTOM = 'Custom',
}

/**
 * Instructor privileges for default instructor roles.
 */
export class DefaultPrivileges {
  /**
   * Instructor privileges for co-owner role.
   */
  static COOWNER: DefaultPrivileges = new DefaultPrivileges(Role.COOWNER, {
    courseLevel: {
      canmodifycourse: true,
      canmodifyinstructor: true,
      canmodifysession: true,
      canmodifystudent: true,
      canviewstudentinsection: true,
      canviewsessioninsection: true,
      cansubmitsessioninsection: true,
      canmodifysessioncommentinsection: true,
    },
    sectionLevel: {},
    sessionLevel: {},
  });

  /**
   * Instructor privileges for manager role.
   */
  static MANAGER: DefaultPrivileges = new DefaultPrivileges(Role.MANAGER, {
    courseLevel: {
      canmodifycourse: false,
      canmodifyinstructor: true,
      canmodifysession: true,
      canmodifystudent: true,
      canviewstudentinsection: true,
      canviewsessioninsection: true,
      cansubmitsessioninsection: true,
      canmodifysessioncommentinsection: true,
    },
    sectionLevel: {},
    sessionLevel: {},
  });

  /**
   * Instructor privileges for observer role.
   */
  static OBSERVER: DefaultPrivileges = new DefaultPrivileges(Role.OBSERVER, {
    courseLevel: {
      canmodifycourse: false,
      canmodifyinstructor: false,
      canmodifysession: false,
      canmodifystudent: false,
      canviewstudentinsection: true,
      canviewsessioninsection: true,
      cansubmitsessioninsection: false,
      canmodifysessioncommentinsection: false,
    },
    sectionLevel: {},
    sessionLevel: {},
  });

  /**
   * Instructor privileges for tutor role.
   */
  static TUTOR: DefaultPrivileges = new DefaultPrivileges(Role.TUTOR, {
    courseLevel: {
      canmodifycourse: false,
      canmodifyinstructor: false,
      canmodifysession: false,
      canmodifystudent: false,
      canviewstudentinsection: true,
      canviewsessioninsection: true,
      cansubmitsessioninsection: true,
      canmodifysessioncommentinsection: false,
    },
    sectionLevel: {},
    sessionLevel: {},
  });

  /**
   * Instructor privileges for custom role.
   */
  static CUSTOM: DefaultPrivileges = new DefaultPrivileges(Role.CUSTOM, {
    courseLevel: {
      canmodifycourse: false,
      canmodifyinstructor: false,
      canmodifysession: false,
      canmodifystudent: false,
      canviewstudentinsection: false,
      canviewsessioninsection: false,
      cansubmitsessioninsection: false,
      canmodifysessioncommentinsection: false,
    },
    sectionLevel: {},
    sessionLevel: {},
  });

  /**
   * Constructs an instance of a default privilege.
   */
  constructor(private key: string, readonly value: Privileges) {
  }

  /**
   * Returns the 'enum' name.
   */
  toString(): string {
    return this.key;
  }
}

/**
 * The default instructor section level privileges
 */
export const defaultSectionLevelPrivileges: SectionLevelPrivileges = {
  canviewstudentinsection: true,
  canviewsessioninsection: true,
  cansubmitsessioninsection: true,
  canmodifysessioncommentinsection: true,
};
