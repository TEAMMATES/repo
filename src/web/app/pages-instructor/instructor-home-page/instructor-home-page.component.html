<h1>Home</h1>
<div class="card bg-light mb-4">
  <div class="card-body">
    <div class="form-group">
      <div class="input-group">
        <input type="text" class="form-control" [(ngModel)]="studentSearch" (keyup.enter)="search()"
               placeholder="e.g. Charles Shultz, charles@gmail.com">
        <div class="input-group-append">
          <button class="btn" (click)="search()">Search</button>
        </div>
      </div>
    </div>
  </div>
</div>

<ng-template #noCourseMessage class="bg-light-blue">
  New to TEAMMATES? You may wish to have a look at our <a routerLink="/web/front/help/instructor">Getting Started Guide</a>.
  A video tour is also available in our <a routerLink="/web">home page</a>.
</ng-template>

<div class="row longer-height">
  <div class="col-4">
    <a routerLink="/web/instructor/courses" class="btn btn-primary text-white">Add New Course</a>
  </div>
  <div class="col-8">
    <div class="float-right">
      <h5 class="d-inline"><strong> Sort By: </strong></h5>
      <div class="btn-group" data-toggle="buttons">
        <button class="btn btn-light" [disabled]="isSelectedForSorting(SortBy.COURSE_ID)" (click)="sortCoursesBy(SortBy.COURSE_ID)">Course ID</button>
        <button class="btn btn-light" [disabled]="isSelectedForSorting(SortBy.COURSE_NAME)" (click)="sortCoursesBy(SortBy.COURSE_NAME)">Course Name</button>
        <button class="btn btn-light" [disabled]="isSelectedForSorting(SortBy.CREATION_DATE)" (click)="sortCoursesBy(SortBy.CREATION_DATE)">Creation date</button>
      </div>
    </div>
  </div>
</div>

<div *ngIf="courseTabModels.length >0; else noCourseMessage">
  <div class="card" *ngFor="let courseTabModel of courseTabModels; let idx = index">
    <div class="card-header bg-primary text-white course-tab-header" (click)="courseTabModel.isTabExpanded = !courseTabModel.isTabExpanded">
      <div class="row">
        <div class="col-6">
          <b>[{{ courseTabModel.course.courseId }}]: {{ courseTabModel.course.courseName }}</b>
        </div>
        <div class="col-6 text-right">
          <i class="fas fa-chevron-down" *ngIf="!courseTabModel.isTabExpanded"></i>
          <i class="fas fa-chevron-up" *ngIf="courseTabModel.isTabExpanded"></i>
        </div>
      </div>
    </div>
    <div class="card-body padding-0" *ngIf="courseTabModel.isTabExpanded">
      <tm-sessions-table [sessionsTableRowModels]="courseTabModel.sessionsTableRowModels" [sessionsTableRowModelsSortBy]="courseTabModel.sessionsTableRowModelsSortBy"
                         [sessionsTableRowModelsSortOrder]="courseTabModel.sessionsTableRowModelsSortOrder" [courseCandidates]="courseCandidates"
                         [headerColorScheme]="SessionsTableHeaderColorScheme.WHITE"
                         [columnsToShow]="[ SessionsTableColumn.START_DATE, SessionsTableColumn.END_DATE ]"
                         (sortSessionsTableRowModelsEvent)="sortSessionsTableRowModelsEvent(idx, $event)"
                         (loadResponseRateEvent)="loadResponseRateEventHandler(idx, $event)" (editSessionEvent)="editSessionEventHandler(idx, $event)"
                         (moveSessionToRecycleBinEvent)="moveSessionToRecycleBinEventHandler(idx, $event)" (copySessionEvent)="copySessionEventHandler(idx, $event)"
                         (submitSessionAsInstructorEvent)="submitSessionAsInstructorEventHandler(idx, $event)" (viewSessionResultEvent)="viewSessionResultEventHandler()"
                         (publishSessionEvent)="publishSessionEventHandler(idx, $event)" (unpublishSessionEvent)="unpublishSessionEventHandler(idx, $event)"
                         (sendRemindersToStudentsEvent)="sendRemindersToStudentsEventHandler(idx, $event)"></tm-sessions-table>
    </div>
  </div>
</div>