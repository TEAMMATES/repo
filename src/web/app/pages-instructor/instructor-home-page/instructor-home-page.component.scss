.padding-0 {
  padding: 0;
}

.btn-toolbar {
  white-space: nowrap;
  margin-top: 5px;
  margin-bottom: 2px;
  align-items: center;
  justify-content: flex-end;
}

.course-header {
  display: inline-flex;
  align-items: center;
  white-space: nowrap;
}

.dropdown-item {
  border: none;
}

.no-border {
  border: none;
}

.text-color-lightgray {
  color: lightgray;
}

.margin-top-bottom-30px {
  margin-top: 30px;
  margin-bottom: 30px;
}
