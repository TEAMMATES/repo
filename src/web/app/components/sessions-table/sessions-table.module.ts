import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { NgbModule } from '@ng-bootstrap/ng-bootstrap';
import { AjaxLoadingModule } from '../ajax-loading/ajax-loading.module';
import { AjaxPreloadModule } from '../ajax-preload/ajax-preload.module';
import { CopySessionModalModule } from '../copy-session-modal/copy-session-modal.module';
import { TeammatesCommonModule } from '../teammates-common/teammates-common.module';
import { FormatDateBriefPipe } from './format-date-brief.pipe';
import { PublishStatusTooltipPipe } from './publish-status-tooltip.pipe';
import {
  ResendResultsLinkToStudentModalComponent,
} from './resend-results-link-to-student-modal/resend-results-link-to-student-modal.component';
import {
  SendRemindersToStudentModalComponent,
} from './send-reminders-to-student-modal/send-reminders-to-student-modal.component';
import { SessionsTableComponent } from './sessions-table.component';
import { StudentListInfoTableComponent } from './student-list-info-table/student-list-info-table.component';
import { SubmissionStatusTooltipPipe } from './submission-status-tooltip.pipe';

/**
 * Module for sessions table.
 */
@NgModule({
  declarations: [
    SubmissionStatusTooltipPipe,
    PublishStatusTooltipPipe,
    SessionsTableComponent,
    ResendResultsLinkToStudentModalComponent,
    SendRemindersToStudentModalComponent,
    FormatDateBriefPipe,
    StudentListInfoTableComponent,
  ],
  imports: [
    CommonModule,
    AjaxLoadingModule,
    AjaxPreloadModule,
    TeammatesCommonModule,
    NgbModule,
    FormsModule,
    CopySessionModalModule,
  ],
  entryComponents: [
    ResendResultsLinkToStudentModalComponent,
    SendRemindersToStudentModalComponent,
  ],
  exports: [
    SessionsTableComponent,
    StudentListInfoTableComponent,
  ],
})
export class SessionsTableModule { }
