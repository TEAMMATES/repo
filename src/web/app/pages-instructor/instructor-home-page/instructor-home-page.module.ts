import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { SessionsTableModule } from '../../components/sessions-table/sessions-table.module';
import { InstructorHomePageComponent } from './instructor-home-page.component';
import { RouterModule } from "@angular/router";

/**
 * Module for instructor home page.
 */
@NgModule({
  declarations: [
    InstructorHomePageComponent,
  ],
  imports: [
    CommonModule,
    SessionsTableModule,
    FormsModule,
    RouterModule,
  ],
  exports: [
    InstructorHomePageComponent,
  ],
})
export class InstructorHomePageModule { }
