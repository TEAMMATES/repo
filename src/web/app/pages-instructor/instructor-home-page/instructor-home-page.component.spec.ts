import { HttpClientTestingModule } from '@angular/common/http/testing';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { SessionsTableRowModel, SortBy, SortOrder } from '../../components/sessions-table/sessions-table-model';
import { InstructorPrivilege } from '../../instructor-privilege';
import { InstructorHomePageComponent } from './instructor-home-page.component';
import { InstructorHomePageModule } from './instructor-home-page.module';

describe('InstructorHomePageComponent', () => {
  let component: InstructorHomePageComponent;
  let fixture: ComponentFixture<InstructorHomePageComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        InstructorHomePageModule,
        HttpClientTestingModule,
        RouterTestingModule,
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(InstructorHomePageComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should snap with default fields', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should snap with one course without feedback session', () => {
    const instructorName: string = '';
    const courseTabModels: any = {
      course: {
        courseId: 'CS3281',
        courseName: 'Thematic Systems',
        creationDate: '26 Feb 23:59 PM',
        timeZone: 'Asia/Singapore',
      },
      instructorPrivilege: {
        canModifyCourse: true,
        canModifySession: true,
        canModifyStudent: true,
        canSubmitSessionInSections: true,
      },
      sessionsTableRowModels: [],
      sessionsTableRowModelsSortBy: SortBy.NONE,
      sessionsTableRowModelsSortOrder: SortOrder.ASC,
      hasPopulated: true,
      isAjaxSuccess: true,
      isTabExpanded: true,
    };
    component.user = instructorName;
    component.courseTabModels = [courseTabModels];
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
