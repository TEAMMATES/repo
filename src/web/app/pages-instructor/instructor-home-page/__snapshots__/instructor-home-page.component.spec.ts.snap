// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstructorHomePageComponent should snap with default fields 1`] = `
<tm-instructor-home-page
  SessionsTableColumn={[Function Object]}
  SessionsTableHeaderColorScheme={[Function Object]}
  SortBy={[Function Object]}
  courseTabModels={[Function Array]}
  feedbackQuestionsService={[Function FeedbackQuestionsService]}
  feedbackSessionsService={[Function FeedbackSessionsService]}
  httpRequestService={[Function HttpRequestService]}
  instructorCoursesSortBy={[Function Number]}
  navigationService={[Function NavigationService]}
  ngbModal={[Function NgbModal]}
  route={[Function ActivatedRoute]}
  router={[Function Router]}
  statusMessageService={[Function StatusMessageService]}
  studentSearchkey=""
  timezoneService={[Function TimezoneService]}
  user="undefined"
>
  <h1>
    Home
  </h1><div
    class="card bg-light mb-4"
  >
    <div
      class="card-body"
    >
      <div
        class="form-group"
      >
        <div
          class="input-group"
        >
          <input
            class="form-control ng-untouched ng-pristine ng-valid"
            ng-reflect-model=""
            placeholder="e.g. Charles Shultz, charles@gmail.com"
            type="text"
          />
          <div
            class="input-group-append"
          >
            <button
              class="btn btn-primary"
            >
              Search
            </button>
          </div>
        </div>
      </div>
    </div>
  </div><div
    class="row height-60px"
  >
    <div
      class="col-4"
    >
      <a
        class="btn btn-primary text-white"
        href="/web/instructor/courses"
        ng-reflect-router-link="/web/instructor/courses"
        routerlink="/web/instructor/courses"
      >
        Add New Course
      </a>
    </div>
    <div
      class="col-8"
    >
      <div
        class="float-right"
      >
        <h5
          class="d-inline"
        >
          <strong>
             Sort By: 
          </strong>
        </h5>
        <div
          class="btn-group"
          data-toggle="buttons"
        >
          <button
            class="btn btn-light"
          >
            Course ID
          </button>
          <button
            class="btn btn-light"
          >
            Course Name
          </button>
          <button
            class="btn btn-light"
            disabled=""
          >
            Creation Date
          </button>
        </div>
      </div>
    </div>
  </div> New to TEAMMATES? You may wish to have a look at our <a
    href="/web/front/help/instructor"
    ng-reflect-router-link="/web/front/help/instructor"
    routerlink="/web/front/help/instructor"
  >
    Getting Started Guide
  </a>. A video tour is also available in our <a
    href="/web"
    ng-reflect-router-link="/web"
    routerlink="/web"
  >
    home page
  </a>.
  
</tm-instructor-home-page>
`;

exports[`InstructorHomePageComponent should snap with one course without feedback session 1`] = `
<tm-instructor-home-page
  SessionsTableColumn={[Function Object]}
  SessionsTableHeaderColorScheme={[Function Object]}
  SortBy={[Function Object]}
  courseTabModels={[Function Array]}
  feedbackQuestionsService={[Function FeedbackQuestionsService]}
  feedbackSessionsService={[Function FeedbackSessionsService]}
  httpRequestService={[Function HttpRequestService]}
  instructorCoursesSortBy={[Function Number]}
  navigationService={[Function NavigationService]}
  ngbModal={[Function NgbModal]}
  route={[Function ActivatedRoute]}
  router={[Function Router]}
  statusMessageService={[Function StatusMessageService]}
  studentSearchkey=""
  timezoneService={[Function TimezoneService]}
  user=""
>
  <h1>
    Home
  </h1><div
    class="card bg-light mb-4"
  >
    <div
      class="card-body"
    >
      <div
        class="form-group"
      >
        <div
          class="input-group"
        >
          <input
            class="form-control ng-untouched ng-pristine ng-valid"
            ng-reflect-model=""
            placeholder="e.g. Charles Shultz, charles@gmail.com"
            type="text"
          />
          <div
            class="input-group-append"
          >
            <button
              class="btn btn-primary"
            >
              Search
            </button>
          </div>
        </div>
      </div>
    </div>
  </div><div
    class="row height-60px"
  >
    <div
      class="col-4"
    >
      <a
        class="btn btn-primary text-white"
        href="/web/instructor/courses"
        ng-reflect-router-link="/web/instructor/courses"
        routerlink="/web/instructor/courses"
      >
        Add New Course
      </a>
    </div>
    <div
      class="col-8"
    >
      <div
        class="float-right"
      >
        <h5
          class="d-inline"
        >
          <strong>
             Sort By: 
          </strong>
        </h5>
        <div
          class="btn-group"
          data-toggle="buttons"
        >
          <button
            class="btn btn-light"
          >
            Course ID
          </button>
          <button
            class="btn btn-light"
          >
            Course Name
          </button>
          <button
            class="btn btn-light"
            disabled=""
          >
            Creation Date
          </button>
        </div>
      </div>
    </div>
  </div><div>
    
    <div
      class="card margin-bottom-15px"
    >
      <div
        class="card-header bg-primary text-white course-tab-header"
      >
        <div
          class="row"
        >
          <div
            class="col-6"
          >
            <b>
              [CS3281]: Thematic Systems
            </b>
          </div>
          <div
            class="col-sm-6"
          >
            
            <div
              class="float-right"
            >
              <span
                class="padding-right-10px dropdown"
                ngbdropdown=""
              >
                <button
                  aria-expanded="false"
                  aria-haspopup="true"
                  class="btn btn-secondary btn-sm dropdown-toggle"
                  ngbdropdowntoggle=""
                >
                   Students 
                </button>
                <div
                  class="dropdown-menu"
                  ngbdropdownmenu=""
                  x-placement="bottom-left"
                >
                  
                  
                  <a
                    class="btn btn-info btn-sm dropdown-item"
                    href="/web/instructor/courses/enroll?courseid=CS3281"
                    ng-reflect-ngb-tooltip="Enroll student into the course"
                    ng-reflect-query-params="[object Object]"
                    ng-reflect-router-link="/web/instructor/courses/enroll"
                    ngbtooltip="Enroll student into the course"
                    routerlink="/web/instructor/courses/enroll"
                  >
                     Enroll 
                  </a>
                  
                  <a
                    class="btn btn-info btn-sm dropdown-item"
                    href="/web/instructor/courses/details?courseid=CS3281"
                    ng-reflect-ngb-tooltip="View, edit and send invitation"
                    ng-reflect-query-params="[object Object]"
                    ng-reflect-router-link="/web/instructor/courses/detail"
                    ngbtooltip="View, edit and send invitation emails to the students in the course"
                    routerlink="/web/instructor/courses/details"
                  >
                     View / Edit 
                  </a>
                </div>
              </span>
              <span
                class="padding-right-10px dropdown"
                ngbdropdown=""
              >
                <button
                  aria-expanded="false"
                  aria-haspopup="true"
                  class="btn btn-secondary btn-sm dropdown-toggle"
                  ngbdropdowntoggle=""
                >
                   Instructors 
                </button>
                <div
                  class="dropdown-menu"
                  ngbdropdownmenu=""
                  x-placement="bottom-left"
                >
                  <a
                    class="btn btn-light btn-sm dropdown-item"
                    href="/web/instructor/courses/edit?courseid=CS3281"
                    ng-reflect-ngb-tooltip="Edit course information and in"
                    ng-reflect-query-params="[object Object]"
                    ng-reflect-router-link="/web/instructor/courses/edit"
                    ngbtooltip="Edit course information and instructor list"
                    routerlink="/web/instructor/courses/edit"
                  >
                     View / Edit 
                  </a>
                </div>
              </span>
              <span
                class="padding-right-10px dropdown"
                ngbdropdown=""
              >
                <button
                  aria-expanded="false"
                  aria-haspopup="true"
                  class="btn btn-secondary btn-sm dropdown-toggle"
                  ngbdropdowntoggle=""
                >
                   Sessions 
                </button>
                <div
                  class="dropdown-menu"
                  ngbdropdownmenu=""
                  x-placement="bottom-left"
                >
                  <a
                    class="btn btn-light btn-sm dropdown-item"
                    href="/web/instructor/sessions?courseid=CS3281"
                    ng-reflect-ngb-tooltip="Add a feedback session for the"
                    ng-reflect-query-params="[object Object]"
                    ng-reflect-router-link="/web/instructor/sessions"
                    ngbtooltip="Add a feedback session for the course"
                    routerlink="/web/instructor/sessions"
                  >
                     Add 
                  </a>
                </div>
              </span>
              <span
                class="padding-right-10px dropdown"
                ngbdropdown=""
              >
                <button
                  aria-expanded="false"
                  aria-haspopup="true"
                  class="btn btn-secondary btn-sm dropdown-toggle"
                  ngbdropdowntoggle=""
                >
                   Course 
                </button>
                <div
                  class="dropdown-menu"
                  ngbdropdownmenu=""
                  x-placement="bottom-left"
                >
                  <a
                    class="btn btn-light btn-sm dropdown-item"
                    ng-reflect-ngb-tooltip="Archive the course so that it "
                    ngbtooltip="Archive the course so that it will not be shown in the home page any more (you can still access it from the 'Courses' tab)"
                  >
                     Archive 
                  </a>
                  <a
                    class="btn btn-light btn-sm dropdown-item"
                    href="/web/instructor/courses/edit?courseid=CS3281"
                    ng-reflect-ngb-tooltip="Edit course information and in"
                    ng-reflect-query-params="[object Object]"
                    ng-reflect-router-link="/web/instructor/courses/edit"
                    ngbtooltip="Edit course information and instructor list"
                    routerlink="/web/instructor/courses/edit"
                  >
                     View / Edit 
                  </a>
                  
                  
                  <a
                    class="btn btn-light btn-sm dropdown-item"
                    ng-reflect-ngb-tooltip="Delete the course and its corr"
                    ngbtooltip="Delete the course and its corresponding students and sessions"
                  >
                     Delete 
                  </a>
                  
                </div>
              </span>
              
              
              <i
                class="fas fa-chevron-up"
              />
            </div>
          </div>
        </div>
      </div>
      
      
      
      
      <div
        class="card-body padding-0"
      >
        <tm-sessions-table
          ng-reflect-columns-to-show="1,2"
          ng-reflect-course-candidates="[object Object]"
          ng-reflect-header-color-scheme="1"
          ng-reflect-sessions-table-row-models=""
          ng-reflect-sessions-table-row-models-sort-by="0"
          ng-reflect-sessions-table-row-models-sort-order="1"
        >
          <div
            class="row"
          >
            <div
              class="col-12"
            >
              <table
                class="table table-responsive-lg table-striped table-bordered margin-bottom-0"
              >
                <thead>
                  <tr
                    ng-reflect-ng-class="[object Object]"
                  >
                    
                    <th
                      class="sortable-header"
                    >
                       Session Name 
                      
                      <span
                        class="text-secondary"
                        ng-reflect-ng-class="[object Object]"
                      >
                        <i
                          class="fas fa-sort"
                        />
                      </span>
                      
                      
                    </th>
                    
                    <th
                      class="sortable-header"
                    >
                       Start Date 
                      
                      <span
                        class="text-secondary"
                        ng-reflect-ng-class="[object Object]"
                      >
                        <i
                          class="fas fa-sort"
                        />
                      </span>
                      
                      
                    </th>
                    
                    <th
                      class="sortable-header"
                    >
                       End Date 
                      
                      <span
                        class="text-secondary"
                        ng-reflect-ng-class="[object Object]"
                      >
                        <i
                          class="fas fa-sort"
                        />
                      </span>
                      
                      
                    </th>
                    <th>
                      Submissions
                    </th>
                    <th>
                      Responses
                    </th>
                    <th
                      container="body"
                      ng-reflect-container="body"
                      ng-reflect-ngb-tooltip="Number of students submitted /"
                      ngbtooltip="Number of students submitted / Class size"
                    >
                      <span
                        class="tool-tip-decorate"
                      >
                        Response Rate
                      </span>
                    </th>
                    <th>
                      Action(s)
                    </th>
                  </tr>
                </thead>
                <tbody>
                  
                </tbody>
              </table>
            </div>
          </div>
        </tm-sessions-table>
      </div>
    </div>
  </div>
</tm-instructor-home-page>
`;
